DxDispatch/src/model/Model.h

```diff
@@ -21,16 +21,17 @@ class Model
     struct BufferBindingSource
     {
         std::string name;
-        uint64_t elementCount;
-        uint64_t elementSizeInBytes;
-        uint64_t elementOffset;
+        uint64_t elementCount  = 0;
+        uint64_t elementSizeInBytes = 0;
+        uint64_t elementOffset = 0;
         std::optional<DXGI_FORMAT> format;
 
         // For Append/Consume buffers only:
         std::optional<std::string> counterName;
-        uint64_t counterOffsetBytes;
+        uint64_t counterOffsetBytes = 0;
 
         std::vector<int64_t> shape;
+        bool deferredBinding = false;
```

Comment:
I agree it is not ideal.

BufferBindingSource is needed so that the onnxDispatchable knows to populate the tensor.  

I would rather get  rid of specifying deferred in the resources section of the json config although I believe that will take some other updates

Reply:
You still need a binding source, but you don't have to indicate that it's deferred. Example:

```json
    "resources": 
    {
        "a": 
        {
            "initialValuesDataType": "FLOAT32",
            "initialValues": [ 1, 2, 3, 4, 5, 6 ]
        },
        "b": 
        {
            "initialValuesDataType": "FLOAT32",
            "initialValues": [ 7, 8, 9, 10, 11, 12 ]
        },
        "c": 
        {
            "initialValuesDataType": "FLOAT32",
            "initialValues": "deferred"
        }
    },

    "dispatchables": 
    {
        "concat": 
        {
            "type": "onnx",
            "sourcePath": "onnx_dynamic_shapes.onnx"
        }
    },

    "commands": 
    [
        {
            "type": "dispatch",
            "dispatchable": "concat",
            "bindings": 
            {
                "x0": { "name": "a", "shape": [1,2] },
                "x1": { "name": "b", "shape": [1,2] },
                "y": { "name": "c" }
            }
        },
        {
            "type": "print",
            "resource": "c"
        }
    ]
```

Reply:
ok so I will keep the variable in the struct BufferBindingSource so that OnnxDispatchable knows to overwrite the values but I will remove the option to specify deferred in the binding json

Reply:
(minor naming) If I saw the name "deferredBinding" by itself in code (without seeing the definition), I would think that "deferredBinding" is some kind of binding, but it's not actually a `DML_BINDING_DESC`, just a boolean that says to *use* deferred binding. Propose `useDeferredBinding` / `wantDeferredBinding` / `hasDeferredBinding` ...

Reply:
After some more thought, I'm open to having a `Model::BufferDesc::initialShape` so that you can explicitly set the shape of a resource (for printing/writing to file) or have it implicitly set by a dispatch (in the case of deferred resource). However, I do think it's useful to retain `BufferBindingSource` (which should override any shape in the resource) and not merge the two. 

I didn't want to resources to impose an abstraction that is higher level than D3D12, which separates the idea of memory (resources) and views.